src/goto-programs/goto_program.cpp

```diff
@@ -748,6 +748,18 @@ void goto_programt::instructiont::validate(
       "asserting non-boolean condition\n" + guard.pretty(),
       source_location);
     break;
+  case FUNCTION_CALL:
+    DATA_CHECK_WITH_DIAGNOSTICS(
+      code.get_statement() == ID_function_call,
+      "function call instruction should contain a call statement",
+      source_location);
+    break;
+  case RETURN:
```

Comment:
Well if not having a return is valid then I guess we shouldn't report it unless it's only allowed for a certain type of goto programs. In this PR we only check present instructions: if something is absent, it would pass undetected.

Reply:
Apologies; I probably wasn't very clear.  There are several (hopefully nested but maybe not) ideas of what a "valid" goto-program is that are implicitly assumed by different parts of the code.  Part of the point of all of this validity checking was to map out, clarify and correct these.  Parts of the code assume that there are no RETURN instructions, parts assume that there are and things like the last instruction (apart from END_INSTRUCTION) should be return.  I was asking how your validity checks will cope with this.  It seems there are two options:

1. The caller specifies which kind of goto-program they want to check for.
2. The check reports some auxiliary flags and the caller then checks these separately.

I don't have particular opinions on how this should be done but I think there needs to be a way of doing so as otherwise... this will miss a lot of the potential benefits.

Reply:
Does a function with removed returns still contain a return statement? This seems to be asserting that return statements have the right shape, not that they must be present or absent.

Reply:
@hannes-steffenhagen-diffblue : No; they are removed.

Reply:
As above, structural properties will be in a different PR.

Reply:
OK.